Balance endpoint integration

paweljakubas opened the pull request:
<!--
Detail in a few bullet points the work accomplished in this PR.

Before you submit, don't forget to:

* Make sure the GitHub PR fields are correct:
   ✓ Set a good Title for your PR.
   ✓ Assign yourself to the PR.
   ✓ Assign one or more reviewer(s).
   ✓ Link to a Jira issue, and/or other GitHub issues or PRs.
   ✓ In the PR description delete any empty sections
     and all text commented in <!--, so that this text does not appear
     in merge commit messages.

* Don't waste reviewers' time:
   ✓ If it's a draft, select the Create Draft PR option.
   ✓ Self-review your changes to make sure nothing unexpected slipped through.

* Try to make your intent clear:
   ✓ Write a good Description that explains what this PR is meant to do.
   ✓ Jira will detect and link to this PR once created, but you can also
     link this PR in the description of the corresponding Jira ticket.
   ✓ Highlight what Testing you have done.
   ✓ Acknowledge any changes required to the Documentation.
-->
adp-656

We have preliminary impl of balanceTransaction that needs to work:
1. coin selection with external inputs (@jonathanknowles ) DONE
2. updateTx impl in shelley (@Anviking ) DONE

running tests:
```
stack test --ta '-m "TRANS_NEW_BALANCE_"' cardano-wallet:integration
```

### Comments

<!-- Additional comments, links, or screenshots to attach, if any. -->

### Issue Number

<!-- Reference the Jira/GitHub issue that this PR relates to, and which requirements it tackles.
  Note: Jira issues of the form ADP- will be auto-linked. -->

paweljakubas committed:
obtain execution prices from network

paweljakubas committed:
add scaffolding for balanceTransaction to be filled

paweljakubas committed:
torwards using selectAssets in balanceTransaction

paweljakubas committed:
change updateTx, more work on impl of balanceTransaction

paweljakubas committed:
work on txCtx

paweljakubas committed:
use ttl and outputs from decodedTx

hlint

paweljakubas committed:
use properly coin selection with external inputs

paweljakubas committed:
enable updating with collaterals

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -2193,12 +2197,116 @@ constructTransaction ctx genChange (ApiT wid) body = do
+                UTxOIndex.fromSequence externalInputs
+
+        let utxoAvailable = UTxOSelection.fromIndexPair
+                (internalUtxoAvailable, externalSelectedUtxo)
```
This looks good. 👍🏻

paweljakubas committed:
unpend the test

Anviking commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -2193,12 +2197,116 @@ constructTransaction ctx genChange (ApiT wid) body = do
+
+    let UnsignedTx colls inps _outs change _wdrl = unsignedtx
+    let extraBody = ExtraTxBodyContent
+            { extraInputs = (\(txin, _, _) -> txin) <$> F.toList inps
```
Won't `inps` already include the pre-selected inputs from `sealedTxIncoming`?

`updateTx` would then try to append the whole `inps` to the pre-selected inputs.

Although as the ledger represents the inputs as a `Set`, it might not actually matter in the end.

paweljakubas replied:
yes, we may filter out preselected inputs. although ledger stores inputs as set so it should not affect txbody or fee.

Anviking committed:
Filter out exteral addresses from ApiCoinSelection

For pre-selected external addresses, we cannot find derivation paths. So
rather than erroring:

```
"selectionToUnsignedTx: unable to find derivation path of a known \
         \input or change address. This is impossible.")
        (traverse withDerivationPath hasAddresses)
```
This commit filters them out.

This has the effect of:
1. Making `stack test --ta '-m "TRANS_NEW_BALANCE_"'
   cardano-wallet:integration pass`
2. Removing the requirement that all ApiCoinSelections have a non-empty
   list of inputs.

paweljakubas committed:
enable test 1d - normal tx with missing inputs

some cleaning

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -2193,12 +2197,116 @@ constructTransaction ctx genChange (ApiT wid) body = do
+            if Map.member acct wdrlMap then
+                mkRewardAccountBuilder @_ @s @_ @n ctx wid (Just SelfWithdrawal)
+            else
+                mkRewardAccountBuilder @_ @s @_ @n ctx wid Nothing
```
```suggestion
        (wdrl, _) <- mkRewardAccountBuilder @_ @s @_ @n ctx wid $
            if Map.member acct wdrlMap
            then Just SelfWithdrawal
            else Nothing
```

paweljakubas replied:
done

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Types/Tx.hs:
```diff
@@ -426,11 +426,14 @@ data UnsignedTx input output change withdrawal = UnsignedTx
         -- transaction body makes it seemingly unique).
+        --
+        -- *However* when used to respresent the inputs known by the wallet, in
+        -- contrast to all inputs, it can be empty.
```
```suggestion
        -- *However* when used to represent the inputs known by the wallet, in
        -- contrast to all inputs, it can be empty.
```

Anviking commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -2193,12 +2197,116 @@ constructTransaction ctx genChange (ApiT wid) body = do
+        let transform s sel' =
+                W.assignChangeAddresses genChange sel' s
+                & uncurry (W.selectionToUnsignedTx (txWithdrawal txCtx))
+        unsignedtx <- liftHandler $ W.selectAssets @_ @s @k wrk W.SelectAssetsParams
```
Because `selectAssets` only knows about the things we tell it, and not the entire `sealedTxIncoming`, I think exotic things like submitting update proposals will result in imbalanced txs.

In a re-thought approach, related to ADP-891, and I think which you've mentioned, we could calculate `minfee sealedTxIncoming` instead and go from there (either by overestimating for the TxExtraBodyContent, or trying to find some fix point)

But I guess the similar and more pressing issue is datums, which everyone was aware of on today's call.

paweljakubas replied:
agree here, it is reasonable to start with minFee, and then increment knowing the cost per inp/out/coll. But we need to remember that evaluation of min fee requires number of witnesses, so it would influence selecting inputs/collaterals in feedback loop. It is not staightforward for me how, in details, it should work

Anviking commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1346,28 +1346,24 @@ selectionToUnsignedTx wdrl sel s =
             fullyQualifiedWithdrawal wdrl
         }
   where
+    -- NOTE: External addresses, not known to the wallet, will be filtered out.
```
When talking briefly with Nikola and hearing / being reminded that they will present the result of balanceTransaction in full detail to the user, this seems wrong.

We probably _should_:
1. Make the `DerivationPath` optional on `ApiCoinSelection`
2. Don't filter out pre-selected inputs/outputs
3. Ensure all other possible tx data is exposed. (We're probably missing glue code for showing datums in api here)

Anviking replied:
(Not critical for merging this PR and I could have a look)

paweljakubas replied:
Daedalus will invoke balance tx, so it will know details of external inputs already - see "inputs" in https://input-output-hk.github.io/cardano-wallet/api/edge/#operation/balanceTransaction And in `ApiCoinSelection` we will present ADDED inps/outs/colls (always being the resource of the wallet). So the client will know everything having "inputs" and response. 
Some argument in favor of what is now is that it is in line what constructTx responds, and as both construct and balance are first step before sign+submit. It makes sense to be compatible....

paweljakubas committed:
some cleanings

some cleaning

Anviking commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -2193,12 +2197,115 @@ constructTransaction ctx genChange (ApiT wid) body = do
+        (ApiTxOut (ApiT addr, _) _ (Quantity amt) (ApiT assets))) =
+        (TxIn hashTx ix, TxOut addr (TokenBundle (Coin $ fromIntegral amt) assets))
+    externalInputs = toTxInTxOut <$> apiExternalInps
+    sealedTxIncoming = body ^. #transaction . #getApiT
```
Just some stylistic thoughts:
I think I have more of a taste for either:
1. Defining important values as `let` bindings at the top
2. Making the `where` definitions all functions rather than values

such that you can read it from top to bottom and easily see where `txIncoming` is coming from.

Anviking commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -2193,12 +2197,115 @@ constructTransaction ctx genChange (ApiT wid) body = do
+balanceTransaction ctx genChange (ApiT wid) body = do
+    pp <- liftIO $ NW.currentProtocolParameters nl
+
+    --TODO  deal with coll and validity and delegations
```
👍 Yes.

Collateral we could simply forbid [as discussed](https://input-output-rnd.slack.com/archives/GBT05825V/p1633089400064100).

Delegations and other stuff could be present in the sealedTxIncoming, but are probably not _crucial_. (unused by PAB)

Validity, however, does sound crucial to support. (used by PAB)

Related to https://github.com/input-output-hk/cardano-wallet/pull/2906#discussion_r721160180

Anviking replied:
🤔 might have gotten this slightly wrong. Will revisit tomorrow.

Anviking commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/TransactionsNew.hs:
```diff
@@ -857,8 +855,6 @@ spec = describe "NEW_SHELLEY_TRANSACTIONS" $ do
 
     it "TRANS_NEW_BALANCE_01e - plutus with missing covering inputs wallet enough funds" $ \ctx -> runResourceT $ do
 
-        liftIO $ pendingWith "Needs all prerequisites integrated with balanceTransaction - to be fixed in ADP-656"
```
_Could_ probably have more assertions in the `verify rTx` below, but as you say, the most crucial thing is the full e2e scenarios.

Anviking reviewed (commented):
It is great to see this start coming together concretely!

My take is that remaining problems are:
1. Critical: Ensure fees are not underestimated because of `datums`
2. Critical: Ensure e.g. ProtocolParameter Hash, or other stuff won't lead to underestimated fees
3. Moderate: Is the API response for CoinSelection appropriate for Daedalus?
4. Low/Moderate: Certificates, update proposals leading to underestimated fees

and that we shouldn't _rush_ to merge this before seeing how the critical things will pan out.

I suspect there are 3 potential solutions for (1):
a. Tweak the coin selection in a minimal / hacky way to account for datums
b. Call ledger/node's `minfee` in a loop - first on the partial input tx, and then after adding inputs and outputs until you converge.
c. Call ledger/node's `minfee` once, and then use the existing `TxConstraints` overestimations to add additional inputs and outputs.

## Edit 

I must have gotten confused in the original message. Got things a bit wrong. Now edited. We _are_ adjusting the original binary. Ensuring tx body content is preserved is not the problem. The problem is ensuring the fee estimation also is aware of all that content.
